### DISCUSSION TITLE
`pyserini.search.lucene` recall result not good as `LuceneSearcher.search`

### DISCUSSION THREAD

**Discussion by better629** [castorini/pyserini#1824] [Q&A] [answered]
I have index the corpus using `pyserini.index --collection JsonCollection --generator DefaultLuceneDocumentGenerator --storePositions --storeDocvectors --storeRaw`.

And do a seach using both `pyserini.search.lucene` and `LuceneSearcher.search`, then check the result with the ground-truth documents.

The result is recall of `pyserini.search.lucene` is smaller than `LuceneSearcher.search`. But according to the `https://github.com/castorini/pyserini/blob/master/docs/usage-index.md#building-a-bm25-index-direct-java-implementation`, it seems that the result should be the same. Anyone know why?

**Comment by lintool** [comment]
hi @better629 - how are the other metrics beyond recall? have you examined the output ranked list and manually compared?

**Reply by better629** [reply]
@lintool According to `https://github.com/castorini/pyserini/blob/master/docs/usage-index.md#building-a-bm25-index-direct-java-implementation`, the candidates rank and score should be same. But not in my case.
I only check the recall, due to the ground-truth is not strictly ordered. Yes, check some cases, the ranked list from result is not exactly same, also the score.

So, the question is that the result should `pyserini.search.lucene` be same with `LuceneSearcher.search` using default params? I have checked the source code, seems they are different implement.

**Reply by better629** [reply]
Here is a search example with same query but different method.
`LuceneSearcher.search`  

```
{
    "instance_id": "astropy__astropy-12057",
    "hits": [
        {
            "docid": "astropy/nddata/nduncertainty.py",
            "score": 185.04710388183594
        },
        {
            "docid": "astropy/nddata/ccddata.py",
            "score": 156.85459899902344
        },
        {
            "docid": "astropy/time/core.py",
            "score": 67.38169860839844
        },
        {
            "docid": "astropy/io/fits/column.py",
            "score": 65.2665023803711
        },
        {
            "docid": "astropy/stats/funcs.py",
            "score": 65.12989807128906
        }
    ]
}
```

`pyserini.search.lucene`  

```
{
    "instance_id": "astropy__astropy-12057",
    "hits": [
        {
            "docid": "astropy/nddata/nduncertainty.py",
            "score": 182.894394
        },
        {
            "docid": "astropy/nddata/ccddata.py",
            "score": 172.855698
        },
        {
            "docid": "astropy/stats/funcs.py",
            "score": 136.316193
        },
        {
            "docid": "astropy/io/fits/column.py",
            "score": 135.997406
        },
        {
            "docid": "astropy/extern/ply/yacc.py",
            "score": 129.202393
        }
    ]
}
```

**Reply by lintool** [reply]
Yes, different parameters seems like the likely culprit...

**Reply by better629** [reply]
I use default params with both methods. So, how to keep the two results be same ?

**Reply by lintool** [reply]
Why not explicitly set the parameters so they are identical, as opposed to relying on the defaults?

**Reply by better629** [reply]
seems the two implement code is different, so the params maybe different.
Do you have any idea which parameters to set in order to get same results?

**Reply by lintool** [reply]
`k1` and `b` are the relevant parameters for BM25:

+ `pyserini.search.lucene`: https://github.com/castorini/pyserini/blob/master/pyserini/search/lucene/__main__.py#L87
+ `LuceneSearcher.search`: https://github.com/castorini/pyserini/blob/master/pyserini/search/lucene/_searcher.py#L345
+ Defaults for `LuceneSearcher` (which wraps `SimpleSearcher` in Anserini) are defined in https://github.com/castorini/anserini/blob/master/src/main/java/io/anserini/search/SearchCollection.java#L265

**Comment by better629** [comment, marked as answer]
@lintool I found the reason, due to the query string contain `\n \r`, so has a escaping like from `\n` to `\\n` and write to tsv file(if using `\n`, write to tsv failed). And then, the text from `query_iterator` is a little different from raw string. This cause the recall result difference.

After input raw string as a param into `search.lucene`, both got same results.

**Reply by lintool** [reply]
Glad you resolved the issue!
